pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskGenerator.java

```diff
@@ -0,0 +1,461 @@
+        return aStartTime != bStartTime ? Long.compare(aStartTime, bStartTime)
+            : Long.compare(a.getEndTimeMs(), b.getEndTimeMs());
+      });
+      for (OfflineSegmentZKMetadata segment : allSegments) {
+        if (preSelectedSegmentsBasedOnLineage.contains(segment.getSegmentName())) {
+          preSelectedSegments.add(segment);
+        }
+      }
+
+      if (preSelectedSegments.isEmpty()) {
+        LOGGER.info("Skip generating task: {} for table: {}, no segment is found to merge.", taskType,
+            offlineTableName);
+        continue;
+      }
+
+      // Sort merge levels based on bucket time period
+      Map<String, Map<String, String>> mergeLevelToConfigs = MergeRollupTaskUtils.getLevelToConfigMap(taskConfigs);
+      List<Map.Entry<String, Map<String, String>>> sortedMergeLevelConfigs =
+          new ArrayList<>(mergeLevelToConfigs.entrySet());
+      sortedMergeLevelConfigs.sort(Comparator.comparingLong(
+          e -> TimeUtils.convertPeriodToMillis(e.getValue().get(MinionConstants.MergeTask.BUCKET_TIME_PERIOD_KEY))));
+
+      // Get incomplete merge levels
+      Set<String> inCompleteMergeLevels = new HashSet<>();
+      for (Map.Entry<String, TaskState> entry : TaskGeneratorUtils.getIncompleteTasks(taskType, offlineTableName,
+          _clusterInfoAccessor).entrySet()) {
+        for (PinotTaskConfig taskConfig : _clusterInfoAccessor.getTaskConfigs(entry.getKey())) {
+          inCompleteMergeLevels.add(taskConfig.getConfigs().get(MergeRollupTask.MERGE_LEVEL_KEY));
+        }
+      }
+
+      ZNRecord mergeRollupTaskZNRecord = _clusterInfoAccessor.getMinionMergeRollupTaskZNRecord(offlineTableName);
+      int expectedVersion = mergeRollupTaskZNRecord != null ? mergeRollupTaskZNRecord.getVersion() : -1;
+      MergeRollupTaskMetadata mergeRollupTaskMetadata =
+          mergeRollupTaskZNRecord != null ? MergeRollupTaskMetadata.fromZNRecord(mergeRollupTaskZNRecord)
+              : new MergeRollupTaskMetadata(offlineTableName, new TreeMap<>());
+      // Schedule tasks from lowest to highest merge level (e.g. Hourly -> Daily -> Monthly -> Yearly)
+      for (int i = 0; i < sortedMergeLevelConfigs.size(); i++) {
+        String lowerMergeLevel = i > 0 ? sortedMergeLevelConfigs.get(i - 1).getKey() : null;
+        String mergeLevel = sortedMergeLevelConfigs.get(i).getKey();
+        Map<String, String> mergeConfigs = sortedMergeLevelConfigs.get(i).getValue();
+
+        // Skip scheduling if there's incomplete task for current mergeLevel
+        if (inCompleteMergeLevels.contains(mergeLevel)) {
+          LOGGER.info("Found incomplete task of merge level: {} for the same table: {}, Skipping task generation: {}",
+              mergeLevel, offlineTableName, taskType);
+          continue;
+        }
+
+        // Get the bucket size and buffer
+        long bucketMs =
+            TimeUtils.convertPeriodToMillis(mergeConfigs.get(MinionConstants.MergeTask.BUCKET_TIME_PERIOD_KEY));
```

Comment:
What happens if the user somehow did not correctly configure the fields? I think that this line will throw the exception and the entire round of scheduling will stop. I think that we should skip, emit some metric or write logs, and then move on to the next table.

Reply:
Ideally the validation should happened at table creation/update phase?

Reply:
Added validation function `alidateMergeRollupTaskConfig()` in `TableConfigUtils` to make sure the fields are configured and valid.